### PR TITLE
DRAFT Crimson/osd: scrub scheduling

### PR DESCRIPTION
**Pull request by ronen-fr**
No actual scrubbing for now - just scheduling 10 seconds "fake scrubs".

### CODE REVIEW

File: src/crimson/osd/osd_operations/scrub_event.cc [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+          } else {
+            return (*(pg->get_scrubber(ScrubberPasskey{})).*std::get<ScrubEvent::ScrubEventFwdFut>(event_fwd_func))(epoch_queued);
+          }
+  
+        }).then_interruptible([this, pg]() mutable {
+          logger().info("ScrubEvent::start() {} after calling fwder", *this);
+          handle.exit();
+          logger().info("ScrubEvent::start() {} executing... exited", *this);
+          return seastar::now();//return complete_rctx(pg);
+        }).then_interruptible([pg]() -> ScrubEvent::interruptible_future<> {
```
Comment (rzarzynski) [inline src/crimson/osd/osd_operations/scrub_event.cc:176]: `finally()` instead?

File: src/crimson/osd/osd_operations/scrub_event.h [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+  spg_t pgid;
+  epoch_t epoch_queued;
+  std::chrono::milliseconds delay{0s};
+
+  const spg_t get_pgid() const { return pgid; }
+
+  virtual void on_pg_absent();
+  // virtual ScrubEvent::interruptible_future<> complete_rctx(Ref<PG>);
+  virtual seastar::future<> complete_rctx_no_pg() { return seastar::now(); }
+  virtual seastar::future<Ref<PG>> get_pg() final;
```
Comment (rzarzynski) [inline src/crimson/osd/osd_operations/scrub_event.h:101]: If there is `final` or `override`, `virtual` stops bringing any further information to a person who reads the code.

**Review by rzarzynski — COMMENTED** [review]
First batch of comments.

File: src/crimson/osd/osd_operations/scrub_event.cc [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+{
+  logger().warn("{}: pg absent, dropping", *this);
+}
+
+seastar::future<Ref<PG>> ScrubEvent::get_pg()
+{
+  return seastar::make_ready_future<Ref<PG>>(pg);
+}
+
+// ScrubEvent::interruptible_future<> ScrubEvent::complete_rctx(Ref<PG> pg)
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/crimson/osd/osd_operations/scrub_event.cc:104]: sure. I did warn you about this being not fully cleaned :-)

File: src/crimson/osd/osd_operations/scrub_event.h [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+
+
+  friend fmt::formatter<ScrubEvent>;
+};
+
+}  // namespace crimson::osd
+
+template <>
+struct fmt::formatter<crimson::osd::ScrubEvent> {
+  constexpr auto parse(format_parse_context& ctx) { return ctx.begin(); }
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/crimson/osd/osd_operations/scrub_event.h:151]: added.

File: src/crimson/osd/osd_operations/scrub_event.cc [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+		       Scrub::act_token_t tkn,
+		       std::chrono::milliseconds delay)
+    : pg{std::move(pg)}
+    , event_fwd_func{func}
+    , act_token{tkn}
+    , shard_services{shard_services}
+    , pgid{pgid}
+    , epoch_queued{epoch_queued}
+    , delay{delay}
+    , dbg_desc{"<ScrubEvent>"}
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/crimson/osd/osd_operations/scrub_event.cc:40]: Not any more. Will remove.

File: src/crimson/osd/osd_operations/scrub_event.cc [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+  f->dump_stream("pgid") << pgid;
+  f->close_section();
+}
+
+void ScrubEvent::on_pg_absent()
+{
+  logger().warn("{}: pg absent, dropping", *this);
+}
+
+seastar::future<Ref<PG>> ScrubEvent::get_pg()
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/crimson/osd/osd_operations/scrub_event.cc:99]: the get_pg()?

[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/crimson/osd/osd_operations/scrub_event.cc:97]: I am using it in ScrubEvent::start()

File: src/crimson/osd/scrubber/osd_scrub_sched.cc [hunk 7]
```diff
@@ -0,0 +1,893 @@
...
+}
+}  // namespace
+
+// ////////////////////////////////////////////////////////////////////////// //
+// ScrubJob
+
+#define dout_context (cct)
+#define dout_subsys ceph_subsys_osd
+#undef dout_prefix
+#define dout_prefix *_dout << "osd." << whoami << "  "
```
Comment (rzarzynski) [inline src/crimson/osd/scrubber/osd_scrub_sched.cc:34]: Why do we need the `dout` stuff here?

File: src/crimson/osd/scrubber/osd_scrub_sched.h [hunk 8]
```diff
@@ -0,0 +1,533 @@
...
+#include "common/ceph_atomic.h"
+#include "common/ceph_mutex.h"
+#include "osd/osd_types.h"
+#include "osd/scrubber_common.h"
+#include "include/utime_fmt.h"
+#include "osd/osd_types_fmt.h"
+
+#include "utime.h"
+
+#ifdef WITH_SEASTAR
```
Comment (rzarzynski) [inline src/crimson/osd/scrubber/osd_scrub_sched.h:126]: I'm a bit confused. Is there a way to have a file in the `src/crimson` built without this flag set?

File: src/crimson/osd/scrubber/osd_scrub_sched.h [hunk 9]
```diff
@@ -0,0 +1,533 @@
...
+     */
+    scrub_schedule_t schedule;
+
+    /// pg to be scrubbed
+    const spg_t pgid;
+
+    /// the OSD id (for the log)
+    const int whoami;
+
+    ceph::atomic<qu_state_t> state{qu_state_t::not_registered};
```
Comment (rzarzynski) [inline src/crimson/osd/scrubber/osd_scrub_sched.h:224]: This is yet another place that suggests to me the file is targeting both classical and crimson OSDs.

File: src/crimson/osd/scrubber/osd_scrub_sched.h [hunk 10]
```diff
@@ -0,0 +1,533 @@
...
+
+  ScrubQContainer to_scrub;   ///< scrub jobs (i.e. PGs) to scrub
+  ScrubQContainer penalized;  ///< those that failed to reserve remote resources
+  bool restore_penalized{false};
+
+  double daily_loadavg{0.0};
+  double current_loadavg{0.0};
+  double filtrd_loadavg{1.0}; // decaying-mean loadavg
+
+  static inline constexpr auto registered_job = [](const auto& jobref) -> bool {
```
Comment (rzarzynski) [inline src/crimson/osd/scrubber/osd_scrub_sched.h:438]: Does this bring any benefit in comparison to a plain `static` method?

File: src/crimson/osd/scrubber/osd_scrub_sched.h [hunk 11]
```diff
@@ -0,0 +1,533 @@
...
+   * their scheduled time.
+   *
+   * Note that the returned container holds independent refs to the
+   * scrub jobs.
+   */
+  ScrubQContainer collect_ripe_jobs(ScrubQContainer& group, utime_t time_now);
+
+
+  /// scrub resources management lock (guarding scrubs_local & scrubs_remote)
+  mutable ceph::mutex resource_lock =
```
Comment (rzarzynski) [inline src/crimson/osd/scrubber/osd_scrub_sched.h:470]: TODO: we have a dummy implementation for crimson IIRC but might be worth to double check. However, if in the classical world a mutex is necessary, in crimson `seastar::mutex` might be desired (assuming we yiled CPU).

File: src/crimson/osd/scrubber/osd_scrub_sched.h (hunk 8 above)
Comment (ronen-fr) [reply src/crimson/osd/scrubber/osd_scrub_sched.h:126]: I am trying to limit the number of "Crimson-specific" scrub-related files. I was able, for example, to use a common
scrub_common.h file (with a very limited number of ifdefs). With luck, I will be able to create a common version of
part of the other .h files.

File: src/crimson/osd/scrubber/osd_scrub_sched.h (hunk 10 above)
Comment (ronen-fr) [reply src/crimson/osd/scrubber/osd_scrub_sched.h:438]: I'd rather use lambdas for comparators. A bit shorter and cleaner (but it's a matter of taste).

File: src/crimson/osd/scrubber/osd_scrub_sched.h (hunk 9 above)
Comment (ronen-fr) [reply src/crimson/osd/scrubber/osd_scrub_sched.h:224]: yes - should be merged with the classic version in some future

File: src/crimson/osd/osd_operations/scrub_event.cc (hunk 1 above)
Comment (ronen-fr) [reply src/crimson/osd/osd_operations/scrub_event.cc]: RRR todo

File: src/crimson/osd/osd.cc [hunk 12]
```diff
@@ -1469,4 +1490,182 @@ seastar::future<> OSD::prepare_to_stop()
...
+{
+  logger().info("{}: trying to initiate a scrubbing of {}", __func__, pgid);
+
+  // we have a candidate to scrub. We need some PG information to know if
+  // scrubbing is allowed
+
+  Ref<PG> pg = get_pg(pgid);
+  if (!pg) {
+    // the PG was dequeued in the short timespan between creating the candidates
+    // list (collect_ripe_jobs()) and here
```
Comment (rzarzynski) [inline src/crimson/osd/osd.cc:1624]: Would be a way to express this as a machine validated assertion?

File: src/crimson/osd/osd.cc [hunk 13]
```diff
@@ -1469,4 +1490,182 @@ seastar::future<> OSD::prepare_to_stop()
...
+      Scrub::schedule_result_t::preconditions);
+  }
+
+  return pg->sched_scrub_this_pg();
+}
+
+void OSD::resched_all_scrubs()
+{
+  logger().info("{}: start", __func__);
+  auto all_jobs = get_scrub_services().list_registered_jobs();
```
Comment (rzarzynski) [inline src/crimson/osd/osd.cc:1653]: Any business in this temporary?

File: src/crimson/osd/osd.cc [hunk 14]
```diff
@@ -1469,4 +1490,182 @@ seastar::future<> OSD::prepare_to_stop()
...
+  for (auto& e : all_jobs) {
+
```
Comment (rzarzynski) [inline src/crimson/osd/osd.cc:1655]: nit: no need for the newline.

File: src/crimson/osd/osd_operations/scrub_remote_event.h [hunk 15]
```diff
@@ -0,0 +1,154 @@
...
+  // Ref<OSDOp> scrub_op;
+  std::optional<ScrubEvent> scrub_event;
+  ShardServices& shard_services;
+  pg_shard_t from;
+  spg_t pgid;
+  epoch_t epoch_queued;
+  std::chrono::milliseconds delay{0s};
+
+  // fix to get the PGID from the OP message
+  spg_t get_pgid() const { return scrub_op ? pgid : pgid; }
```
Comment (rzarzynski) [inline src/crimson/osd/osd_operations/scrub_remote_event.h:91]: A tautology.

File: src/crimson/osd/osd_operations/scrub_remote_event.cc [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+	   [this, pg]() mutable -> ScrubRemoteEvent::interruptible_future<> {
+	     (void)((*(pg->get_scrubber(ScrubberPasskey{})).*
+		     event_fwd_func)(conn, scrub_op, 0, from));
+
+	     // just for now
+	     return seastar::make_ready_future<>();
+	   },
+	   [](std::exception_ptr) { return seastar::now(); },
+	   pg)
+    .finally([this, opref, pg] { track_event<CompletionEvent>(); });
```
Comment (rzarzynski) [inline src/crimson/osd/osd_operations/scrub_remote_event.cc:137]: No need catching `opref` nor `pg`.

File: src/crimson/osd/osd_operations/scrub_remote_event.cc [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+seastar::future<> ScrubRemoteEvent::with_pg(ShardServices& shard_services,
+					    Ref<PG> pg)
+{
+  logger().debug("{}: {}", "ScrubRemoteEvent::with_pg", *this);
+
+  track_event<StartEvent>();
+  IRef opref = this;
+
+  // create a scrub event based on the parsed request
+  parse_into_event(shard_services, pg);
```
Comment (rzarzynski) [inline src/crimson/osd/osd_operations/scrub_remote_event.cc:125]: How do set `this->pg`?

File: src/crimson/osd/scrubber/osd_scrub_sched.h (hunk 10 above)
Comment (ronen-fr) [reply src/crimson/osd/scrubber/osd_scrub_sched.h:438]: removing the 'inline'

File: src/crimson/osd/osd_operations/scrub_remote_event.h [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+ * via start_pg_operation(), which means they implement with_pg()
+ */
+
+class ScrubRemoteEvent : public PhasedOperationT<ScrubRemoteEvent> {
+ public:
+  static constexpr OperationTypeCode type = OperationTypeCode::scrub_event;
+  static constexpr bool can_create() { return false; }
+
+  template <typename T = void>
+  using interruptible_future = ::crimson::interruptible::
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/crimson/osd/osd_operations/scrub_remote_event.h:57]: Removed. Thanks

File: src/crimson/osd/osd.cc [hunk 19]
```diff
@@ -1469,4 +1490,182 @@ seastar::future<> OSD::prepare_to_stop()
...
+  logger().info("scrub_tick: load_avg={}", cla);
+
+  static int down_counter = 0;
+  if (down_counter > 0) {
+    down_counter--;
+    return;
+  }
+  down_counter = 5; // RRR should be a config!
+  // maybe: local_conf().get_val<int64_t>("osd_scrub_scheduling_period")");
+  static thread_local seastar::semaphore limit(1);
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/crimson/osd/osd.cc:1536]: Agree. A dev-only line.

File: src/crimson/osd/osd.cc [hunk 20]
```diff
@@ -1469,4 +1490,182 @@ seastar::future<> OSD::prepare_to_stop()
...
+    }
+  }
+  logger().warn("scrub_tick(): after pg loop");
+#endif
 }
+
+
+seastar::future<> OSD::sched_scrub()
+{
+  if (!state.is_active()) {
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/crimson/osd/osd.cc:1575]: Not sure. Not being active is OK, and the scrub code is already too chatty

File: src/crimson/osd/osd.cc [hunk 21]
```diff
@@ -1469,4 +1490,182 @@ seastar::future<> OSD::prepare_to_stop()
...
+    return seastar::now();
+  }
+  if (scrub_random_backoff()) {
+    return seastar::now();
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/crimson/osd/osd.cc:1579]: logged in scrub_random_backoff()

File: src/crimson/osd/osd.cc [hunk 22]
```diff
@@ -1469,4 +1490,182 @@ seastar::future<> OSD::prepare_to_stop()
...
+                    ScrubQueue::attempt_res_text(was_started));
+      return seastar::now();
+    });
+}
+
+
+bool OSD::scrub_random_backoff()
+{
+  bool coin_flip =
+    (rand() / (double)RAND_MAX) >= local_conf()->osd_scrub_backoff_ratio;
```
[the remark this answers is not in the document]
Comment (ronen-fr) [reply src/crimson/osd/osd.cc:1632]: Agree - and we should stop using rand(), BTW.
"It is implementation-defined whether rand() is thread-safe"

File: src/crimson/osd/osd.cc (hunk 12 above)
Comment (ronen-fr) [reply src/crimson/osd/osd.cc:1654]: That would be wrong. PG's do come and go.

File: src/crimson/osd/osd_operations/scrub_remote_event.h (hunk 15 above)
Comment (ronen-fr) [reply src/crimson/osd/osd_operations/scrub_remote_event.h:91]: more like a placeholder...
Will fix.

File: src/crimson/osd/osd_operations/scrub_remote_event.cc (hunk 17 above)
Comment (ronen-fr) [reply src/crimson/osd/osd_operations/scrub_remote_event.cc:124]: now set in parse_into_event()

**Pull request closed without merging**